### ISSUE TITLE
[Docker Engine] Add multi-platform image support (migrate to the containerd content store and snapshotters)

### ISSUE THREAD

**Issue by chris-crone** [docker/roadmap#371]
**Tell us about your request**
Docker started the containerd project with the plan for the Docker Engine to be migrated to use it. The Docker Engine already uses containerd to manage container lifecycle but it still uses its own implementation for image and filesystem management.

Docker is working with the Moby project contributors on completing the migration and moving Docker Engine to use containerd's content store and snapshotters. This will enable low-level features like multi-platform image support. It will also reduce the code complexity of the Docker Engine.

**Which service(s) is this request for?**
[Docker Engine](https://github.com/moby/moby) which is a core part of Docker Desktop and Docker for Linux.

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
There are many user facing features that this will enable. To list a few:

* Native multi-platform image build support (i.e.: do not require containerized BuildKit)
* Resource namespacing
* More OCI support
* Support for containerd’s snapshotters

**Comment by sam-thibault** [comment]
The containerd image store feature is now in beta in Docker Desktop 4.12.0 🎉
More info is available here: [https://docs.docker.com/desktop/containerd/](https://docs.docker.com/desktop/containerd/)

We’d love to hear from you! Please feel free to provide us feedback or report any bugs you may find through the issues tracker on the [feedback form](https://dockr.ly/3PODIhD).

**Comment by NiklasBr** [comment]
Building with `--platform` works, but gives this warning:
```
WARNING: No output specified with docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load
```

Using`--load` as suggested by the warning results in this error:
```
ERROR: docker exporter does not currently support exporting manifest lists
```

Can you explain what is supposed to happen here and why following the instructions does not work?

**Comment by NiklasBr** [comment]
> Which instructions were you following?

These instructions:

> ```WARNING: No output specified with docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load```

So, I append `--load` to the command, and then I get this message instead:

> ```ERROR: docker exporter does not currently support exporting manifest lists```

**Comment by TBBle** [comment]
Okay, if you weren't following a specific tutorial (sorry, I misunderstood) then the answer is there in the output: If using the docker-container driver, you must provide either `--push` or `--load` in order to get the built manifest or manifest list out of the build cache (this is true irrespective of the containerd integration, and is as-expected), and the docker exporter (triggered by `--load`) does not handle manifest lists in that version of buildx (which is also not related to the containerd image store; the cause here is that without the conatinerd content store, docker doesn't know how to store manifest lists at all, so the docker exporter didn't support them either).

Basically, you've hit https://github.com/docker/buildx/issues/59#issuecomment-1189999088. I still guess that the fix is to enable both the containerd content store beta (as you are) so that manifest lists can be stored and also the newer version of buildx I mentioned earlier, which knows how to store manifest lists in Docker.

**Comment by jacobwgillespie** [comment]
Does any version of buildx support multi-platform exports with the beta containerd store? My understanding is that exporting an image to `docker` (`--load`) still explicitly blocks exporting multi-platform images:

https://github.com/docker/buildx/blob/fb27e3f919dcbf614d7126b10c2bc2d0b1927eb6/build/build.go#L507-L509

**Comment by TBBle** [comment]
Drat. That's why it was a guess. 🐍👀.

My guess was based on https://github.com/moby/buildkit/pull/3389 having been merged for BuildKit 0.11, and BuildKit 0.11 having been vendored into buildx 0.10. However, on closer inspection, it (and https://github.com/moby/buildkit/pull/3380 which is how I got there) are not solving this issue, but were preventing the same error message being triggered by other changes.

So I guess there's still some work to go in buildx to be able to push a manifest list via the Docker exporter (`--load`) into the content store when containerd is enabled, and perhaps even work needed in the Docker Engine side and/or BuildKit too.

**Comment by TBBle** [comment]
Drat. That's why it was a guess. 🐍👀.

My guess was based on https://github.com/moby/buildkit/pull/3389 having been merged for BuildKit 0.11, and BuildKit 0.11 having been vendored into buildx 0.10. However, on closer inspection, it (and https://github.com/moby/buildkit/pull/3380 which is how I got there) are not solving this issue, but were preventing the same error message being triggered by other changes.

So I guess there's still some work to go in buildx to be able to push a manifest list via the Docker exporter (`--load`) into the content store when containerd is enabled, and perhaps even work needed in the Docker Engine side and/or BuildKit too.

**Comment by TBBle** [comment]
Drat. That's why it was a guess. 🐍👀.

My guess was based on https://github.com/moby/buildkit/pull/3389 having been merged for BuildKit 0.11, and BuildKit 0.11 having been vendored into buildx 0.10. However, on closer inspection, it (and https://github.com/moby/buildkit/pull/3380 which is how I got there) are not solving this issue, but were preventing the same error message being triggered by other changes.

So I guess there's still some work to go in buildx to be able to push a manifest list via the Docker exporter (`--load`) into the content store when containerd is enabled, and perhaps even work needed in the Docker Engine side and/or BuildKit too.

**Comment by TBBle** [comment]
Thought I should probably make some effort on my own behalf.

On a reset-to-defaults and then containerd-beta-activated Docker Desktop for Windows:

```
> docker buildx ls
NAME/NODE       DRIVER/ENDPOINT STATUS  BUILDKIT                         PLATFORMS
default *       docker
  default       default         running 22.06.0-beta.0-913-g6be7129598.m linux/amd64, linux/amd64/v2, linux/amd64/v3, linux/arm64, linux/riscv64, linux/ppc64le, linux/s390x, linux/mips64le, linux/mips64
desktop-linux   docker
  desktop-linux desktop-linux   running 22.06.0-beta.0-913-g6be7129598.m linux/amd64, linux/amd64/v2, linux/amd64/v3, linux/arm64, linux/riscv64, linux/ppc64le, linux/s390x, linux/mips64le, linux/mips64
```

So looks like the `docker` driver is multiplatform now? From this state, the trivial Dockerfile from the documentation worked.

So to recreate the observed issue:

```
docker buildx create --name docker-container --driver docker-container
docker buildx --builder docker-container build --platform linux/amd64,linux/arm64 -t hello-friends .
```
indeed produces
```
WARNING: No output specified with docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load
```

and adding `--load` or `--output=type=docker` produces

```
ERROR: docker exporter does not currently support exporting manifest lists
```

and then trying the workaround:

```
docker buildx --builder docker-container build --output=type=oci,dest= --platform linux/amd64,linux/arm64 -t hello-friends .
```

produces the following error:

```
ERROR: failed to solve: method /moby.filesync.v1.FileSend/diffcopy not supported by the client
```

which is hitting [this code](https://github.com/moby/buildkit/blob/c521afdba5d609c343bcb0ce938d16b361301785/session/filesync/filesync.go#L306-L310) from [here](https://github.com/moby/buildkit/blob/c521afdba5d609c343bcb0ce938d16b361301785/exporter/oci/export.go#L248), as the next expected line was "sending tarball". So either this code-path is not yet implemented, or I've done something that broke it. (I'm not sure which code is "the client" here, my guess is buildx, and maybe it's doing something specifically in `docker` output mode to implement FileSend that isn't implemented in `oci` mode).

**Comment by TBBle** [comment]
> Both `docker buildx --builder default …` and `docker buildx --builder epic_wilbur …` fails with the same error `ERROR: docker exporter does not currently support exporting manifest lists`.

Oh, with `--builder default`, take out `--load`, that's overriding the exporter from `image` to `docker`.

**Comment by TBBle** [comment]
Based on my last batch of testing, I went ahead and opened https://github.com/docker/buildx/issues/1522 for https://github.com/docker/roadmap/issues/371#issuecomment-1380326975 and ensuing discussion, since it _looks like_ all the plumbing is there in BuildKit and Docker for something like `--load` that supports multi-platform images, it's just the glue in buildx that's missing.

**Comment by AaronFriel** [comment]
> Native multi-platform image build support (i.e.: do not require containerized BuildKit)

Would this enable `docker build --platform linux/amd64,linux/arm64,...` (sans buildx) and the Go SDK to support platform lists as well?

i.e.:

```go
	opts := types.ImageBuildOptions{
		Dockerfile: img.Build.Dockerfile,
		Tags:       []string{img.Name},
		Remove:     true,
		CacheFrom:  img.Build.CachedImages,
		BuildArgs:  build.Args,
		Version:    build.BuilderVersion,
		Platform:   "linux/amd64,linux/arm64", // and more

		AuthConfigs: authConfigs,
	}
```

**Comment by dhduvall** [comment]
I tried the `docker buildx build --output=type=oci | docker load` workaround, but I always get `open /var/lib/docker/tmp/docker-import-2070341532/blobs/json: no such file or directory`. The tarball looks entirely reasonable to me. All I've been able to dig up on the message suggests that it happens when the tarball is not the format it expects, and most often happens when people do `docker export | docker load`. Using `docker import` on the tarball doesn't error out, but I can't create a container from the image.

This is Docker Desktop on Mac (4.16.2); engine and client versions are both 20.10.22, `buildx version` is `v0.10.0`. Will the workaround only work with newer versions?

**Comment by TBBle** [comment]
As I noted in https://github.com/docker/buildx/issues/1522, if `--output=type=oci,dest=somefile.tar` and the `docker load -i somefile.tar` works for the multi-platform case, then it might be something weird in the output stream when not writing to a file, as I saw in PowerShell but not Git Bash on Windows.

If that does _not_ work, then something else is wrong, different from Docker Desktop for Windows where I did my testing. I don't know enough about Docker Desktop on Mac to know what the difference might be.

I don't recall trying to execute the resulting image, but I do know that `docker image` has not been updated to handle manifest lists/image indexes in its storage, as it was showing two images that were 100% identical, and it seemed to only act upon one image when asked to, e.g. `docker image inspect`, except `docker image rm` which removed both. I didn't look into where in the data flow this is going wrong, I'm _hoping_ this is just a docker-cli limitation to be addressed.

**Comment by crazy-max** [comment]
> I tried the `docker buildx build --output=type=oci | docker load` workaround, but I always get `open /var/lib/docker/tmp/docker-import-2070341532/blobs/json: no such file or directory`. The tarball looks entirely reasonable to me. All I've been able to dig up on the message suggests that it happens when the tarball is not the format it expects, and most often happens when people do `docker export | docker load`. Using `docker import` on the tarball doesn't error out, but I can't create a container from the image.
> 
> This is Docker Desktop on Mac (4.16.2); engine and client versions are both 20.10.22, `buildx version` is `v0.10.0`. Will the workaround only work with newer versions?
> 
> Edit: I tried exporting just one platform, and that loaded and ran just fine. I have to tag it manually, but it gets the job done.

Since [Buildx 0.10](https://github.com/docker/buildx/releases/tag/v0.10.0), a minimal [SLSA Provenance](https://slsa.dev/provenance/) attestation is attached and therefore creates an [OCI-compliant](https://github.com/opencontainers/image-spec) multi-platform image. For this case you might need to disable provenance with `--provenance=false`.

**Comment by zhixinwen** [comment]
I am compiling x86 on my M2 machine, and I enabled containerd on Docker.
This gives me error `parent snapshot sha256:b6bab4b00958273f1ba360d4fbb20cf6b4ee829508f4b4985e16d51a28b3a0f0 does not exist: not found`

**Comment by AaronFriel** [comment]
The containerd backend for Docker is still considered experimental, so I'm afraid that's likely an issue with Docker for Mac.

**Comment by NiklasBr** [comment]
Any updates here?

**Comment by NiklasBr** [comment]
What's the timeline until we will be able to use multi-platform builds using `docker buildx build --platform=linux/amd64,linux/arm64` it's been a while since you wrote about it at https://www.docker.com/blog/multi-platform-docker-builds/

**Comment by savannahostrowski** [comment]
Containerd image store support is now generally available with the release of Moby 25 (shipped in [DD 4.27](https://docs.docker.com/desktop/release-notes/#4270)). There will be a blog post that covers the DD 4.27 soon. Docs can be found at https://docs.docker.com/desktop/containerd/.

**Comment by NiklasBr** [comment]
Looking forward to additional documentation because the link did not work with containerd and SBOM enabled:

```
❯ docker buildx create --bootstrap

[+] Building 5.5s (1/1) FINISHED                                                                                                                                                                                                                                                      
 => [internal] booting buildkit                                                                                                                                                                                                                                                  5.5s
 => => pulling image moby/buildkit:buildx-stable-1                                                                                                                                                                                                                               4.9s
 => => creating container buildx_buildkit_pedantic_brattain0                                                                                                                                                                                                                     0.6s
pedantic_brattain
❯ docker info --format="{{ .Driver }}"

overlay2
❯ docker buildx build --platform=linux/arm64,linux/amd64 --tag=user/containerd-multiplatform ./docker/proxy-nginx
[+] Building 0.0s (0/0)                                                                                                                                                                                                                                          docker:desktop-linux
ERROR: Multi-platform build is not supported for the docker driver.
```

**Comment by dvdksn** [comment]
hi @NiklasBr. From your `docker info` command it looks like you aren't using the containerd storage backend.

```
docker info --format="{{ .Driver }}"
overlay2
```

`overlay2` is the default graph driver. If you enable the containerd image store feature in Docker Desktop and try again, building multi-platform images and loading them to the local image store should work fine, even with the default buildx driver.

**Comment by TBBle** [comment]
Assuming there were more lines to that error
```
Switch to a different driver, or turn on the containerd image store, and try again.
Learn more at https://docs.docker.com/go/build-multi-platform/
```
then it appears your docker buildx doesn't believe you're using the containerd image store.

If no such lines appeared, then I have no idea what version of buildx you're using, but I can't see any way in the buildx 0.12.1 release to generate that error without those extra lines, and 0.11 and older had different phrasing for that error.

----

I just upgraded my Docker Desktop to 4.27.1, and trivially built the container from https://docs.docker.com/get-started/02_our_app/ using the given command:
```
git clone https://github.com/docker/getting-started-app.git
cd .\getting-started-app
# Populate the Dockerfile from the example
gvim .\Dockerfile
docker buildx build --platform=linux/arm64,linux/amd64 --tag=user/containerd-multiplatform .
```

and it appears to have worked as expected:
```
> docker image list
REPOSITORY                      TAG       IMAGE ID       CREATED              SIZE
user/containerd-multiplatform   latest    03d0489bce84   About a minute ago   86.6MB
user/containerd-multiplatform   latest    03d0489bce84   2 minutes ago        352MB
```
